### DISCUSSION TITLE
Parallelization in pymor

### DISCUSSION THREAD

**Discussion by pdiercks** [pymor/pymor#1764] [Q&A]
Hi all, 

I am a bit confused about how to run pymor code in parallel in general and also in combination with a model that is discretized using fenics(x). My current understanding is that, mainly `pymor.parallel.interface.WorkerPool` (or `pymor.parallel.default.new_parallel_pool` to create a pool) is intended to be used to run e.g. a greedy algorithm in parallel.
Unfortunately, I did not find any other source of information on this in the docs or the demos (apologies, if I overlooked something!).

At the moment, I am experimenting with fenicsx (would also be happy to contribute to the experimental bindings).
I have a pure fenicsx code that runs without problem with the following command:
```sh
mpirun -n 4 python fenicsx_code.py
```
However, if I import anything from pymor, I get stuck in the event loop that is initialized by pymor (`__init__.py`)
```sh
mpirun -n 4 python fenicsx_code.py
Rank 1: MPI parallel run detected. Launching event loop ...
```
and nothing happens for a long time. How do I have to interact with the `event_loop`?

Could someone please explain and help me understanding how to use pymor with `mpi4py`?

Best,
Philipp

**Comment by renefritze** [comment]
You did not really miss anything @pdiercks. We're sorely lacking a tutorial for this stuff. Thanks for reaching out about it!

Maybe have a look at the [fenics_nonlinear demo](https://github.com/pymor/pymor/blob/main/src/pymordemos/fenics_nonlinear.py#L20). 
The [auto wrapping](https://github.com/pymor/pymor/blob/93fe56909f88a4e92bedc27ef4de14edcdc46755/src/pymor/models/mpi.py#L82) of a model will basically give you an object where operations are rank-local and there's communication magic happening on the [vector array](https://github.com/pymor/pymor/blob/main/src/pymor/vectorarrays/mpi.py) level, with results landing on rank 0.

You can turn off the event loop launching by setting the [corresponding default](https://github.com/pymor/pymor/blob/93fe56909f88a4e92bedc27ef4de14edcdc46755/src/pymor/tools/mpi.py#L87). 

Everything else very much depends on what you actually want to do with pyMOR in your setting.

**Reply by sdrave** [reply]
@pdiercks, you are right about the `WorkerPool` stuff. This is used to parallelize pyMOR's algorithms (mainly the greedy algorithm). So you probably do not care about that. When the FOM itself is MPI parallelized, you need to use the wrapper classes in `pymor.models.mpi`, `pymor.operators.mpi` and `pymor.vectorarrays.mpi`, which also rely on the event loop. So you definitely do not want to disable it.

Basically, `MPIVectorArray` and friends work as follows: all MPI wrapper object only live on rank 0 and represent the entire distributed `VectorArray` or `Operator`. When you call an interface method on these objects, the event loop mechanism in `pymor.tools.mpi` is used to instruct all ranks to dispatch into their MPI parallelized backend code. If you want to try to understand whats happening, I suggest you first look at the implementation of `MPIVectorArrayImpl`.

**Reply by pdiercks** [reply]
Thanks @sdrave ! I'll have a look at `MPIVectorArrayImpl` and probably there will be more questions at some point ...
